### PR TITLE
Merge precisemodule into modelsmodule

### LINKED ISSUE #1684: Merge precisemodule with modelsmodule
Currently, the "precise" neuron models are built into a separate module ("precisemodule") from the regular neuron and synapse models (which are in "modelsmodule"). During the post-conference 2020 NEST Hackathon, it was suggested to merge the precise models in with the rest of the models.

Action points:
- [ ] Move files; update paths; fix the build
- [ ] Update model documentation for precise models along the lines of #1530 (@sarakonradi)
   - [ ] Fix keywords (@jougs)

### PR DESCRIPTION
**Pull request by jougs**
This PR moves all models from the `precise` directory to the `models` directory and `slice_ring_buffer.*` to `nestkernel`. It also updates pertaining documentation and other pieces of text to reflect this change (fixes #1684).

On top of this, all user-level documentation of the precise models has been converted to RST to be discoverable by the new doc extractor (this is a follow-up to #1530)

### CODE REVIEW

**Comment by jougs** [comment]
@clinssen, @sarakonradi: As I did not get an answer by you, I just went ahead and converted all user-level documentation. Be quick to review this, or I'll add more to it :smirk_cat:

File: models/iaf_psc_alpha_canon.h [hunk 1]
```diff
@@ -71,35 +76,38 @@ This implementation is more complex than the plain iaf_psc_alpha
 neuron, but achieves much higher precision. In particular, it does not
 suffer any binning of spike times to grid points. Depending on your
 application, the canonical application may provide superior overall
-performance given an accuracy goal; see [1] for details.  Subthreshold
-dynamics are integrated using exact integration between events [2].
-
-Remarks:
+performance given an accuracy goal; see [1]_ for details.  Subthreshold
```
Comment (sarakonradi) [inline models/iaf_psc_alpha_canon.h:79]:
```suggestion
performance given an accuracy goal; see [1]_ for details. Subthreshold
```

File: models/iaf_psc_alpha_canon.h [hunk 2]
```diff
@@ -115,31 +123,43 @@ if tau_m is equal to tau_syn, to avoid numerical instabilities.
 For details, please see doc/model_details/IAF_neurons_singularity.ipynb.
 
 A further improvement of precise simulation is implemented in iaf_psc_exp_ps
-based on [3].
+based on [3]_.
 
 For details about exact subthreshold integration, please see
-``doc/model_details/exact-integration.ipynb``.
+<https://github.com/nest/nest-simulator/blob/master/doc/model_details/exact-integration.ipynb>`_
+in the NEST source code (docs/model_details).
```
Comment (sarakonradi) [inline models/iaf_psc_alpha_canon.h:130]:
```suggestion
in the NEST source code (doc/model_details).
```

File: models/iaf_psc_alpha_ps.h [hunk 3]
```diff
@@ -107,28 +113,39 @@ if tau_m is equal to tau_syn_ex/in, to avoid numerical instabilities.
 For details, please see doc/model_details/IAF_neurons_singularity.ipynb.
 
 For details about exact subthreshold integration, please see
-``doc/model_details/exact-integration.ipynb``.
+<https://github.com/nest/nest-simulator/blob/master/doc/model_details/exact-integration.ipynb>`_
+in the NEST source code (docs/model_details).
```
Comment (sarakonradi) [inline models/iaf_psc_alpha_ps.h:117]:
```suggestion
in the NEST source code (doc/model_details).
```

File: models/iaf_psc_exp_ps.h [hunk 4]
```diff
@@ -101,27 +109,38 @@ tau_syn_in, respectively, to avoid numerical instabilities.
 For details, please see doc/model_details/IAF_neurons_singularity.ipynb.
 
 For details about exact subthreshold integration, please see
-``doc/model_details/exact-integration.ipynb``.
+<https://github.com/nest/nest-simulator/blob/master/doc/model_details/exact-integration.ipynb>`_
+in the NEST source code (docs/model_details).
```
Comment (sarakonradi) [inline models/iaf_psc_exp_ps.h:113]:
```suggestion
in the NEST source code (doc/model_details).
```

File: models/iaf_psc_exp_ps_lossless.h [hunk 5]
```diff
@@ -93,22 +104,34 @@ Support for different time constants may be added in the future,
 see issue #921.
 
 For details about exact subthreshold integration, please see
-``doc/model_details/exact-integration.ipynb``.
+<https://github.com/nest/nest-simulator/blob/master/doc/model_details/exact-integration.ipynb>`_
+in the NEST source code (docs/model_details).
```
Comment (sarakonradi) [inline models/iaf_psc_exp_ps_lossless.h:108]:
```suggestion
in the NEST source code (doc/model_details).
```

File: models/iaf_psc_delta_ps.h [hunk 6]
```diff
@@ -123,32 +121,42 @@ CurrentEvents. These events transmit stepwise-constant currents which
 can only change at on-grid times.
 
 For details about exact subthreshold integration, please see
-``doc/model_details/exact-integration.ipynb``.
+<https://github.com/nest/nest-simulator/blob/master/doc/model_details/exact-integration.ipynb>`_
+in the NEST source code (docs/model_details).
```
Comment (sarakonradi) [inline models/iaf_psc_delta_ps.h:125]:
```suggestion
in the NEST source code (doc/model_details).
```

File: models/iaf_psc_alpha_canon.h (hunk 2 above)
Comment (jougs) [reply models/iaf_psc_alpha_canon.h:130]: Actually this link looks weird in the rendered version. I just copied the one from [models/iaf_psc_alpha.h#L95](https://github.com/nest/nest-simulator/blob/master/models/iaf_psc_alpha.h#L95), but that might already be looking strange. can you please have a look?

File: models/iaf_psc_alpha_canon.h [hunk 7]
```diff
@@ -115,31 +123,43 @@ if tau_m is equal to tau_syn, to avoid numerical instabilities.
 For details, please see doc/model_details/IAF_neurons_singularity.ipynb.
```
Comment (sarakonradi) [inline models/iaf_psc_alpha_canon.h:123]:
```suggestion
For details, please check out the `IAF neurons singularity
<https://github.com/nest/nest-simulator/blob/master/doc/model_details/IAF_neurons_singularity.ipynb>`_ notebook.
```

File: models/iaf_psc_alpha_canon.h [hunk 8]
```diff
@@ -115,31 +123,43 @@ if tau_m is equal to tau_syn, to avoid numerical instabilities.
...
 
 A further improvement of precise simulation is implemented in iaf_psc_exp_ps
-based on [3].
+based on [3]_.
 
 For details about exact subthreshold integration, please see
-``doc/model_details/exact-integration.ipynb``.
+<https://github.com/nest/nest-simulator/blob/master/doc/model_details/exact-integration.ipynb>`_
+in the NEST source code (doc/model_details).
```
Comment (sarakonradi) [inline models/iaf_psc_alpha_canon.h:130]:
```suggestion
:doc:`../guides/exact-integration`.
```

File: models/iaf_psc_alpha_ps.h [hunk 9]
```diff
@@ -107,28 +113,39 @@ if tau_m is equal to tau_syn_ex/in, to avoid numerical instabilities.
 For details, please see doc/model_details/IAF_neurons_singularity.ipynb.
```
Comment (sarakonradi) [inline models/iaf_psc_alpha_ps.h:113]:
```suggestion
For details, please check out the `IAF neurons singularity
<https://github.com/nest/nest-simulator/blob/master/doc/model_details/IAF_neurons_singularity.ipynb>`_ notebook.
```

File: models/iaf_psc_alpha_ps.h [hunk 10]
```diff
@@ -107,28 +113,39 @@ if tau_m is equal to tau_syn_ex/in, to avoid numerical instabilities.
...
 
 For details about exact subthreshold integration, please see
-``doc/model_details/exact-integration.ipynb``.
+<https://github.com/nest/nest-simulator/blob/master/doc/model_details/exact-integration.ipynb>`_
+in the NEST source code (doc/model_details).
```
Comment (sarakonradi) [inline models/iaf_psc_alpha_ps.h:117]:
```suggestion
:doc:`../guides/exact-integration`.
```

File: models/iaf_psc_delta_ps.h [hunk 11]
```diff
@@ -123,32 +121,42 @@ CurrentEvents. These events transmit stepwise-constant currents which
 can only change at on-grid times.
 
 For details about exact subthreshold integration, please see
-``doc/model_details/exact-integration.ipynb``.
+<https://github.com/nest/nest-simulator/blob/master/doc/model_details/exact-integration.ipynb>`_
+in the NEST source code (doc/model_details).
```
Comment (sarakonradi) [inline models/iaf_psc_delta_ps.h:125]:
```suggestion
:doc:`../guides/exact-integration`.
```

File: models/iaf_psc_exp_ps.h [hunk 12]
```diff
@@ -101,27 +109,38 @@ tau_syn_in, respectively, to avoid numerical instabilities.
 For details, please see doc/model_details/IAF_neurons_singularity.ipynb.
```
Comment (sarakonradi) [inline models/iaf_psc_exp_ps.h:109]:
```suggestion
For details, please check out the `IAF neurons singularity
<https://github.com/nest/nest-simulator/blob/master/doc/model_details/IAF_neurons_singularity.ipynb>`_ notebook.
```

File: models/iaf_psc_exp_ps.h [hunk 13]
```diff
@@ -101,27 +109,38 @@ tau_syn_in, respectively, to avoid numerical instabilities.
...
 
 For details about exact subthreshold integration, please see
-``doc/model_details/exact-integration.ipynb``.
+<https://github.com/nest/nest-simulator/blob/master/doc/model_details/exact-integration.ipynb>`_
+in the NEST source code (doc/model_details).
```
Comment (sarakonradi) [inline models/iaf_psc_exp_ps.h:113]:
```suggestion
:doc:`../guides/exact-integration`.
```

File: models/iaf_psc_exp_ps_lossless.h [hunk 14]
```diff
@@ -93,22 +104,34 @@ Support for different time constants may be added in the future,
 see issue #921.
 
 For details about exact subthreshold integration, please see
-``doc/model_details/exact-integration.ipynb``.
+<https://github.com/nest/nest-simulator/blob/master/doc/model_details/exact-integration.ipynb>`_
+in the NEST source code (doc/model_details).
```
Comment (sarakonradi) [inline models/iaf_psc_exp_ps_lossless.h:108]:
```suggestion
:doc:`../guides/exact-integration`.
```

**Comment by jougs** [comment]
@sarakonradi: Thanks for the help. I've committed all of your suggestions.

**Pull request merged**